components/core/src/clp/clp/FileDecompressor.cpp

```diff
@@ -76,4 +118,125 @@ bool FileDecompressor::decompress_file(
+        string const& output_dir,
+        string const& temp_output_dir,
+        streaming_archive::reader::Archive& archive_reader,
+        size_t ir_target_size
+) {
+    // Open encoded file
+    auto error_code = archive_reader.open_file(m_encoded_file, file_metadata_ix);
+    if (ErrorCode_Success != error_code) {
+        if (ErrorCode_errno == error_code) {
+            SPDLOG_ERROR("Failed to open encoded file, errno={}", errno);
+        } else {
+            SPDLOG_ERROR("Failed to open encoded file, error_code={}", error_code);
+        }
+        return false;
+    }
+
+    // Generate output directory
+    error_code = create_directory_structure(output_dir, 0700);
+    if (ErrorCode_Success != error_code) {
+        SPDLOG_ERROR(
+                "Failed to create directory structure {}, errno={}",
+                output_dir.c_str(),
+                errno
+        );
+        return false;
+    }
+
+    if (temp_output_dir != output_dir) {
+        // Generate temporary output directory
+        error_code = create_directory_structure(temp_output_dir, 0700);
+        if (ErrorCode_Success != error_code) {
+            SPDLOG_ERROR(
+                    "Failed to create directory structure {}, errno={}",
+                    output_dir.c_str(),
+                    errno
+            );
+            return false;
+        }
+    }
+
+    boost::filesystem::path temp_ir_path{temp_output_dir};
+    temp_ir_path /= m_encoded_file.get_id_as_string() + ".temp." + ir::get_ir_extension_name();
+
+    auto const& file_orig_id = m_encoded_file.get_orig_file_id_as_string();
+    auto begin_message_ix = m_encoded_file.get_begin_message_ix();
+
+    LogEventSerializer<four_byte_encoded_variable_t> ir_serializer;
+    // Open output IR file
+    error_code = ir_serializer.open(temp_ir_path.string(), m_encoded_file.get_begin_ts());
+    if (ErrorCode_Success != error_code) {
+        return false;
+    }
+
+    while (archive_reader.get_next_message(m_encoded_file, m_encoded_message)) {
+        if (!archive_reader
+                     .decompress_message_without_ts(m_encoded_message, m_decompressed_message))
+        {
+            SPDLOG_ERROR("Failed to decompress message");
+            return false;
+        }
+        auto const message_size_as_ir = ir::get_approximated_ir_size(
+                m_decompressed_message,
+                m_encoded_message.get_vars().size()
+        );
+        if (message_size_as_ir + ir_serializer.get_serialized_size() > ir_target_size) {
+            ir_serializer.close();
+
+            auto end_message_ix = begin_message_ix + ir_serializer.get_num_log_events();
+            if (false
+                == rename_and_move_ir(
+                        temp_ir_path,
+                        output_dir,
+                        file_orig_id,
+                        begin_message_ix,
+                        end_message_ix
+                ))
+            {
+                return false;
```

Comment:
A general question: when failure happens, shall we close the serializer before return?

Reply:
technically we should, however, we are not closing archive or metadata either. another issue I can see: if a serializing a log event fails, the IR will be incompletely anyway.

For this specific use case, since we are only decompressing a single file, I don't think closing or not closing makes a difference.
If one day we are going to decompress multiple IRs in a single execution, we can further optimize this